Compilation in C++ fails when -Wpedantic and -Werror=unused-parameter flags are used

I am compiling the library with relaxed flags but I have stricter flags on my C++ code. And because my code includes quickjs.h I get the errors below

```
_deps/quickjs-ng-src/quickjs.h: In function ‘JSValue JS_NewBool(JSContext*, int)’:
_deps/quickjs-ng-src/quickjs.h:181:68: error: ISO C++ forbids compound-literals [-Wpedantic]
  181 | #define JS_MKVAL(tag, val) (JSValue){ (JSValueUnion){ .int32 = val }, tag }
      |                                                                    ^
_deps/quickjs-ng-src/quickjs.h:483:12: note: in expansion of macro ‘JS_MKVAL’
  483 |     return JS_MKVAL(JS_TAG_BOOL, (val != 0));
      |            ^~~~~~~~
_deps/quickjs-ng-src/quickjs.h:181:75: error: ISO C++ forbids compound-literals [-Wpedantic]
  181 | #define JS_MKVAL(tag, val) (JSValue){ (JSValueUnion){ .int32 = val }, tag }
      |                                                                           ^
_deps/quickjs-ng-src/quickjs.h:483:12: note: in expansion of macro ‘JS_MKVAL’
  483 |     return JS_MKVAL(JS_TAG_BOOL, (val != 0));
      |            ^~~~~~~~
_deps/quickjs-ng-src/quickjs.h:481:54: error: unused parameter ‘ctx’ [-Werror=unused-parameter]
  481 | static js_force_inline JSValue JS_NewBool(JSContext *ctx, JS_BOOL val)
      |                                           ~~~~~~~~~~~^~~
_deps/quickjs-ng-src/quickjs.h: In function ‘JSValue JS_NewInt32(JSContext*, int32_t)’:
_deps/quickjs-ng-src/quickjs.h:181:68: error: ISO C++ forbids compound-literals [-Wpedantic]
  181 | #define JS_MKVAL(tag, val) (JSValue){ (JSValueUnion){ .int32 = val }, tag }
      |                                                                    ^
_deps/quickjs-ng-src/quickjs.h:488:12: note: in expansion of macro ‘JS_MKVAL’
  488 |     return JS_MKVAL(JS_TAG_INT, val);
      |            ^~~~~~~~
_deps/quickjs-ng-src/quickjs.h:181:75: error: ISO C++ forbids compound-literals [-Wpedantic]
  181 | #define JS_MKVAL(tag, val) (JSValue){ (JSValueUnion){ .int32 = val }, tag }
      |                                                                           ^
_deps/quickjs-ng-src/quickjs.h:488:12: note: in expansion of macro ‘JS_MKVAL’
  488 |     return JS_MKVAL(JS_TAG_INT, val);
      |            ^~~~~~~~
_deps/quickjs-ng-src/quickjs.h:486:55: error: unused parameter ‘ctx’ [-Werror=unused-parameter]
  486 | static js_force_inline JSValue JS_NewInt32(JSContext *ctx, int32_t val)
      |                                            ~~~~~~~~~~~^~~
_deps/quickjs-ng-src/quickjs.h: In function ‘JSValue JS_NewFloat64(JSContext*, double)’:
_deps/quickjs-ng-src/quickjs.h:491:57: error: unused parameter ‘ctx’ [-Werror=unused-parameter]
  491 | static js_force_inline JSValue JS_NewFloat64(JSContext *ctx, double val)
      |                                              ~~~~~~~~~~~^~~
_deps/quickjs-ng-src/quickjs.h: In function ‘JSValue JS_NewCatchOffset(JSContext*, int32_t)’:
_deps/quickjs-ng-src/quickjs.h:181:68: error: ISO C++ forbids compound-literals [-Wpedantic]
  181 | #define JS_MKVAL(tag, val) (JSValue){ (JSValueUnion){ .int32 = val }, tag }
      |                                                                    ^
_deps/quickjs-ng-src/quickjs.h:498:12: note: in expansion of macro ‘JS_MKVAL’
  498 |     return JS_MKVAL(JS_TAG_CATCH_OFFSET, val);
      |            ^~~~~~~~
_deps/quickjs-ng-src/quickjs.h:181:75: error: ISO C++ forbids compound-literals [-Wpedantic]
  181 | #define JS_MKVAL(tag, val) (JSValue){ (JSValueUnion){ .int32 = val }, tag }
      |                                                                           ^
_deps/quickjs-ng-src/quickjs.h:498:12: note: in expansion of macro ‘JS_MKVAL’
  498 |     return JS_MKVAL(JS_TAG_CATCH_OFFSET, val);
      |            ^~~~~~~~
_deps/quickjs-ng-src/quickjs.h:496:61: error: unused parameter ‘ctx’ [-Werror=unused-parameter]
  496 | static js_force_inline JSValue JS_NewCatchOffset(JSContext *ctx, int32_t val)
      |                                                  ~~~~~~~~~~~^~~
_deps/quickjs-ng-src/quickjs.h: In function ‘int JS_IsBigInt(JSContext*, JSValue)’:
_deps/quickjs-ng-src/quickjs.h:533:46: error: unused parameter ‘ctx’ [-Werror=unused-parameter]
  533 | static inline JS_BOOL JS_IsBigInt(JSContext *ctx, JSValue v)
      |                                   ~~~~~~~~~~~^~~
_deps/quickjs-ng-src/quickjs.h: In function ‘JSValue JS_DupValue(JSContext*, JSValue)’:
_deps/quickjs-ng-src/quickjs.h:614:46: error: unused parameter ‘ctx’ [-Werror=unused-parameter]
  614 | static inline JSValue JS_DupValue(JSContext *ctx, JSValue v)
      |                                   ~~~~~~~~~~~^~~
_deps/quickjs-ng-src/quickjs.h: In function ‘JSValue JS_DupValueRT(JSRuntime*, JSValue)’:
_deps/quickjs-ng-src/quickjs.h:623:48: error: unused parameter ‘rt’ [-Werror=unused-parameter]
  623 | static inline JSValue JS_DupValueRT(JSRuntime *rt, JSValue v)
```

@wuqi yes, this compiles and works but I think the issue is more complex. I tried to use one of the other macros to test it and I now get a different error message:

```
_deps/quickjs-ng-src/quickjs.h:173:43: error: use of old-style cast to ‘int32_t’ {aka ‘int’} [-Werror=old-style-cast]
  173 | #define JS_VALUE_GET_TAG(v) ((int32_t)(v).tag)
      |                                           ^~~
```

So I may need to handle this in a different way

--- a/Makefile
+++ b/Makefile
@@ -83,8 +83,14 @@ distclean:
 stats: $(QJS)
 	$(QJS) -qd
 
-# implicitly .PHONY because it doesn't generate output
-cxxtest: CXXFLAGS+=-std=c++11 -fsyntax-only -Wall -Wextra -Werror -Wno-unused-parameter
+# effectively .PHONY because it doesn't generate output
+ctest: CFLAGS=-std=c11 -fsyntax-only -Wall -Wextra -Werror -pedantic
+ctest: ctest.c quickjs.h
+	$(CC) $(CFLAGS) -DJS_NAN_BOXING=0 $<
+	$(CC) $(CFLAGS) -DJS_NAN_BOXING=1 $<
+
+# effectively .PHONY because it doesn't generate output
+cxxtest: CXXFLAGS=-std=c++11 -fsyntax-only -Wall -Wextra -Werror -pedantic
 cxxtest: cxxtest.cc quickjs.h
 	$(CXX) $(CXXFLAGS) -DJS_NAN_BOXING=0 $<
 	$(CXX) $(CXXFLAGS) -DJS_NAN_BOXING=1 $<
@@ -116,4 +122,4 @@ unicode_gen: $(BUILD_DIR)
 libunicode-table.h: unicode_gen
 	$(BUILD_DIR)/unicode_gen unicode $@
 
-.PHONY: all cxxtest debug fuzz install clean codegen distclean stats test test262 test262-update test262-check microbench unicode_gen $(QJS) $(QJSC)
+.PHONY: all ctest cxxtest debug fuzz install clean codegen distclean stats test test262 test262-update test262-check microbench unicode_gen $(QJS) $(QJSC)

--- a/ctest.c
+++ b/ctest.c
@@ -1,0 +1,17 @@
+// note: file is not actually compiled, only checked for C syntax errors
+#include "quickjs.h"
+
+int main(void)
+{
+    JSRuntime *rt = JS_NewRuntime();
+    JSContext *ctx = JS_NewContext(rt);
+    JS_FreeValue(ctx, JS_NAN);
+    JS_FreeValue(ctx, JS_UNDEFINED);
+    JS_FreeValue(ctx, JS_NewFloat64(ctx, 42));
+    // not a legal way of using JS_MKPTR but this is here
+    // to have the compiler syntax-check its definition
+    JS_FreeValue(ctx, JS_MKPTR(JS_TAG_UNINITIALIZED, 0));
+    JS_FreeContext(ctx);
+    JS_FreeRuntime(rt);
+    return 0;
+}

--- a/cxxtest.cc
+++ b/cxxtest.cc
@@ -1,17 +1,2 @@
 // note: file is not actually compiled, only checked for C++ syntax errors
-#include "quickjs.h"
-
-int main(void)
-{
-    JSRuntime *rt = JS_NewRuntime();
-    JSContext *ctx = JS_NewContext(rt);
-    JS_FreeValue(ctx, JS_NAN);
-    JS_FreeValue(ctx, JS_UNDEFINED);
-    JS_FreeValue(ctx, JS_NewFloat64(ctx, 42));
-    // not a legal way of using JS_MKPTR but this is here
-    // to have the compiler syntax-check its definition
-    JS_FreeValue(ctx, JS_MKPTR(JS_TAG_UNINITIALIZED, 0));
-    JS_FreeContext(ctx);
-    JS_FreeRuntime(rt);
-    return 0;
-}
+#include "ctest.c"

--- a/quickjs.h
+++ b/quickjs.h
@@ -514,21 +514,25 @@ JS_EXTERN int JS_IsRegisteredClass(JSRuntime *rt, JSClassID class_id);
 
 static js_force_inline JSValue JS_NewBool(JSContext *ctx, JS_BOOL val)
 {
+    (void)&ctx;
     return JS_MKVAL(JS_TAG_BOOL, (val != 0));
 }
 
 static js_force_inline JSValue JS_NewInt32(JSContext *ctx, int32_t val)
 {
+    (void)&ctx;
     return JS_MKVAL(JS_TAG_INT, val);
 }
 
 static js_force_inline JSValue JS_NewFloat64(JSContext *ctx, double val)
 {
+    (void)&ctx;
     return __JS_NewFloat64(val);
 }
 
 static js_force_inline JSValue JS_NewCatchOffset(JSContext *ctx, int32_t val)
 {
+    (void)&ctx;
     return JS_MKVAL(JS_TAG_CATCH_OFFSET, val);
 }
 
@@ -566,8 +570,8 @@ static inline JS_BOOL JS_IsNumber(JSValue v)
 
 static inline JS_BOOL JS_IsBigInt(JSContext *ctx, JSValue v)
 {
-    int tag = JS_VALUE_GET_TAG(v);
-    return tag == JS_TAG_BIG_INT;
+    (void)&ctx;
+    return JS_VALUE_GET_TAG(v) == JS_TAG_BIG_INT;
 }
 
 static inline JS_BOOL JS_IsBool(JSValue v)

